WebDriverAgentLib/Utilities/FBMjpegServer.m

```diff
@@ -9,74 +9,128 @@
 
 @import CocoaAsyncSocket;
 
+#import "FBMjpegServer.h"
+
 #import "FBApplication.h"
+#import "FBLogger.h"
 #import "FBMathUtils.h"
-#import "FBMjpegServer.h"
-#import "XCUIDevice+FBHelpers.h"
+#import "XCTestManager_ManagerInterface-Protocol.h"
+#import "FBXCTestDaemonsProxy.h"
+
+static const NSUInteger FPS = 10;
+static const NSTimeInterval SCREENSHOT_TIMEOUT = 0.5;
+static const double SCREENSHOT_QUALITY = 0.25;
 
-static const NSTimeInterval FPS = 10;
 static NSString *const SERVER_NAME = @"WDA MJPEG Server";
+static const char *QUEUE_NAME = "JPEG Screenshots Provider Queue";
+
 
 @interface FBMjpegServer()
 
-@property (nonatomic) NSTimer *mainTimer;
+@property (nonatomic, nullable) NSTimer *mainTimer;
 @property (nonatomic) dispatch_queue_t backgroundQueue;
 @property (nonatomic) NSMutableArray<GCDAsyncSocket *> *activeClients;
-@property (atomic) CGRect screenRect;
+@property (nonatomic, nullable) NSValue *screenRect;
 
 @end
 
-@implementation FBMjpegServer
 
+@implementation FBMjpegServer
 
 - (instancetype)init
 {
   if ((self = [super init])) {
     _activeClients = [NSMutableArray array];
-    _backgroundQueue = dispatch_queue_create("Background screenshoting", DISPATCH_QUEUE_SERIAL);
+    _screenRect = nil;
+    _backgroundQueue = dispatch_queue_create(QUEUE_NAME, DISPATCH_QUEUE_SERIAL);
+    if (![self.class canStreamScreenshots]) {
+      [FBLogger log:@"MJPEG server cannot start because the current iOS version is not supoprted"];
+      return self;
+    }
     _mainTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 / FPS repeats:YES block:^(NSTimer * _Nonnull timer) {
       @synchronized (self.activeClients) {
         if (0 == self.activeClients.count) {
           return;
         }
       }
 
-      if (CGRectIsEmpty(self.screenRect)) {
-        return;
-      }
-      NSData *screenshotData = [[XCUIDevice sharedDevice] fb_rawScreenshotWithQuality:2 rect:self.screenRect error:nil];
-      if (nil == screenshotData) {
-        return;
-      }
+      @synchronized(self.screenRect) {
```

Comment:
space between `@synchronized` and `(` like above?

Reply:
fixed. Actually, there is no need to sync on this property since the stuff is performed on the main thread